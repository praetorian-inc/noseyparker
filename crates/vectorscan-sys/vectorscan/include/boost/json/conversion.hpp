//
// Copyright (c) 2022 Dmitry Arkhipov (grisumbras@yandex.ru)
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
//
// Official repository: https://github.com/boostorg/json
//

#ifndef BOOST_JSON_CONVERSION_HPP
#define BOOST_JSON_CONVERSION_HPP

#include <boost/json/detail/config.hpp>

#include <type_traits>

BOOST_JSON_NS_BEGIN

/** Customization point tag.

    This tag type is used by the function
    @ref value_from to select overloads
    of `tag_invoke`.

    @note This type is empty; it has no members.

    @see @ref value_from, @ref value_to, @ref value_to_tag,
    <a href="http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1895r0.pdf">
        tag_invoke: A general pattern for supporting customisable functions</a>
*/
struct value_from_tag { };

/** Customization point tag type.

    This tag type is used by the function
    @ref value_to to select overloads
    of `tag_invoke`.

    @note This type is empty; it has no members.

    @see @ref value_from, @ref value_from_tag, @ref value_to,
    <a href="http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1895r0.pdf">
        tag_invoke: A general pattern for supporting customisable functions</a>
*/
template<class T>
struct value_to_tag { };

/** Customization point tag type.

    This tag type is used by the function
    @ref try_value_to to select overloads
    of `tag_invoke`.

    @note This type is empty; it has no members.

    @see @ref value_to, @ref value_to_tag
    <a href="http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1895r0.pdf">
        tag_invoke: A general pattern for supporting customisable functions</a>
*/
template<class T>
struct try_value_to_tag { };

/** Determine if `T` can be treated like a string during conversions.

    Provides the member constant `value` that is equal to `true`, if `T` is
    convertible to @ref string_view. Otherwise, `value` is equal to `false`.
    <br>

    Users can specialize the trait for their own types if they don't want them
    to be treated like strings. For example:

    @code
    namespace boost {
    namespace json {

    template <>
    struct is_string_like<your::string> : std::false_type
    { };

    } // namespace boost
    } // namespace json
    @endcode

    @par Types satisfying the trait

    @ref string,
    @ref string_view,
    <a href="https://en.cppreference.com/w/cpp/string/basic_string"><tt>std::string</tt></a>,
    <a href="https://en.cppreference.com/w/cpp/string/basic_string_view"><tt>std::string_view</tt></a>.

    @see @ref value_from, @ref value_to
*/
template<class T>
struct is_string_like;

/** Determine if `T` can be treated like a sequence during conversions.

    Given `t`, a glvalue of type `T`, if

    @li given `It`, the type denoted by `decltype(std::begin(t))`,
        <tt>std::iterator_traits<It>::iterator_category</tt> is well-formed and
        denotes a type; and

    @li `decltype(std::end(t))` also denotes the type `It`;

    then the trait provides the member constant `value` that is equal to
    `true`. Otherwise, `value` is equal to `false`.<br>

    Users can specialize the trait for their own types if they don't want them
    to be treated like sequences. For example:

    @code
    namespace boost {
    namespace json {

    template <>
    struct is_sequence_like<your::container> : std::false_type
    { };

    } // namespace boost
    } // namespace json
    @endcode


    @par Types satisfying the trait

    Any <a href="https://en.cppreference.com/w/cpp/named_req/SequenceContainer"><em>SequenceContainer</em></a>,
    array types.

    @see @ref value_from, @ref value_to
*/
template<class T>
struct is_sequence_like;

/** Determine if `T` can be treated like a 1-to-1 mapping during
    conversions.

    Given `t`, a glvalue of type `T`, if

    @li <tt>is_sequence_like<T>::value</tt> is `true`; and

    @li given type `It` denoting `decltype(std::begin(t))`, and types `K`
        and `M`,  <tt>std::iterator_traits<It>::value_type</tt> denotes
        `std::pair<K, M>`; and

    @li <tt>std::is_string_like<K>::value</tt> is `true`; and

    @li given `v`, a glvalue of type `V`, and `E`, the type denoted by
        `decltype(t.emplace(v))`,
        <tt>std::is_tuple_like<E>::value</tt> is `true`;

    then the trait provides the member constant `value`
    that is equal to `true`. Otherwise, `value` is equal to `false`.<br>

    Users can specialize the trait for their own types if they don't want them
    to be treated like mappings. For example:

    @code
    namespace boost {
    namespace json {

    template <>
    struct is_map_like<your::map> : std::false_type
    { };

    } // namespace boost
    } // namespace json
    @endcode


    @note

    The restriction for `t.emplace()` return type ensures that the container
    does not accept duplicate keys.

    @par Types satisfying the trait

    <a href="https://en.cppreference.com/w/cpp/container/map"><tt>std::map</tt></a>,
    <a href="https://en.cppreference.com/w/cpp/container/unordered_map"><tt>std::unordered_map</tt></a>.

    @see @ref value_from, @ref value_to
*/
template<class T>
struct is_map_like;

/** Determine if `T` can be treated like a tuple during conversions.

    Provides the member constant `value` that is equal to `true`, if
    <tt>std::tuple_size<T>::value</tt> is a positive number. Otherwise, `value`
    is equal to `false`.<br>

    Users can specialize the trait for their own types if they don't want them
    to be treated like tuples. For example:

    @code
    namespace boost {
    namespace json {

    template <>
    struct is_tuple_like<your::tuple> : std::false_type
    { };

    } // namespace boost
    } // namespace json
    @endcode


    @par Types satisfying the trait

    <a href="https://en.cppreference.com/w/cpp/utility/tuple"><tt>std::tuple</tt></a>,
    <a href="https://en.cppreference.com/w/cpp/utility/pair"><tt>std::pair</tt></a>.

    @see @ref value_from, @ref value_to
*/
template<class T>
struct is_tuple_like;

/** Determine if `T` can be treated like null during conversions.

    Primary template instantiations provide the member constant `value` that is
    equal to `false`. Users can specialize the trait for their own types if
    they **do** want them to be treated as nulls. For example:

    @code
    namespace boost {
    namespace json {

    template <>
    struct is_null_like<your::null_type> : std::true_type
    { };

    } // namespace boost
    } // namespace json
    @endcode


    @par Types satisfying the trait

    <a href="https://en.cppreference.com/w/cpp/types/nullptr_t"><tt>std::nullptr_t</tt></a>.

    @see @ref value_from, @ref value_to
*/
template<class T>
struct is_null_like
    : std::false_type
{ };

/** Determine if `T` should be treated as a described class

    Described classes are serialised as objects with an element for each
    described public data member. A described class should not have described
    bases or non-public members.<br>

    Or more formally, given `L`, a class template
    of the form `template<class...> struct L {};`, if

    @li <tt>boost::describe::has_members<T, boost::describe::mod_public>::value</tt> is `true`; and

    @li `boost::describe::describe_members<T, boost::describe::mod_private | boost::describe::mod_protected>` denotes `L<>`; and

    @li `boost::describe::describe_bases<T, boost::describe::mod_any_access>` denotes `L<>`; and

    @li <tt>std::is_union<T>::value</tt> is `false`;

    then the trait provides the member constant `value`
    that is equal to `true`. Otherwise, `value` is equal to `false`.<br>

    Users can specialize the trait for their own types if they don't want them
    to be treated as described classes. For example:

    @code
    namespace boost {
    namespace json {

    template <>
    struct is_described_class<your::described_class> : std::false_type
    { };

    } // namespace boost
    } // namespace json
    @endcode

    Users can also specialize the trait for their own types _with_ described
    bases to enable this conversion implementation. In this case the class will
    be serialized in a flattened way, that is members of bases will be
    serialized as direct elements of the object, and no nested objects will be
    created for bases.

    @see <a href="https://www.boost.org/doc/libs/develop/libs/describe/doc/html/describe.html">Boost.Describe</a>.
*/
template<class T>
struct is_described_class;

/** Determine if `T` should be treated as a described enum

    Described enums are serialised as strings when their value equals to a
    described enumerator, and as integers otherwise. The reverse operation
    does not convert numbers to enums values, though, and instead produces
    an error.<br>

    If <tt>boost::describe::has_describe_enumerators<T>::value</tt> is `true`,
    then the trait provides the member constant `value`
    that is equal to `true`. Otherwise, `value` is equal to `false`.<br>

    Users can specialize the trait for their own enums if they don't want them
    to be treated as described enums. For example:

    @code
    namespace boost {
    namespace json {

    template <>
    struct is_described_enum<your::described_enum> : std::false_type
    { };

    } // namespace boost
    } // namespace json
    @endcode

    @see <a href="https://www.boost.org/doc/libs/develop/libs/describe/doc/html/describe.html">Boost.Describe</a>.
*/
template<class T>
struct is_described_enum;

BOOST_JSON_NS_END

#include <boost/json/impl/conversion.hpp>

#endif // BOOST_JSON_CONVERSION_HPP
