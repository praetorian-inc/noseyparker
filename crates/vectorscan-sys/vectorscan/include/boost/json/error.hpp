//
// Copyright (c) 2019 Vinnie Falco (vinnie.falco@gmail.com)
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
//
// Official repository: https://github.com/boostorg/json
//

#ifndef BOOST_JSON_ERROR_HPP
#define BOOST_JSON_ERROR_HPP

#include <boost/json/detail/config.hpp>
#include <boost/json/system_error.hpp>

BOOST_JSON_NS_BEGIN

/** Error codes returned by JSON operations

*/
enum class error
{
    //
    // parse errors
    //

    /// syntax error
    syntax = 1,

    /// extra data
    extra_data,

    /// incomplete JSON
    incomplete,

    /// exponent too large
    exponent_overflow,

    /// too deep
    too_deep,

    /// illegal leading surrogate
    illegal_leading_surrogate,

    /// illegal trailing surrogate
    illegal_trailing_surrogate,

    /// expected hex digit
    expected_hex_digit,

    /// expected utf16 escape
    expected_utf16_escape,

    /// An object contains too many elements
    object_too_large,

    /// An array contains too many elements
    array_too_large,

    /// A key is too large
    key_too_large,

    /// A string is too large
    string_too_large,

    /// error occured when trying to read input
    input_error,

    //
    // generic errors
    //

    /// An exception was thrown during operation
    exception,

    /// test failure
    test_failure,

    //
    // JSON Pointer errors
    //

    /// missing slash character before token reference
    missing_slash,

    /// invalid escape sequence
    invalid_escape,

    /// token should be a number but cannot be parsed as such
    token_not_number,

    /// current value is neither an object nor an array
    value_is_scalar,

    /// current value does not contain referenced value
    not_found,

    /// token cannot be represented by std::size_t
    token_overflow,

    /// past-the-end index is not supported
    past_the_end,

    //
    // Conversion errors
    //

    /// JSON number was expected during conversion
    not_number,

    /// number cast is not exact
    not_exact,

    /// JSON null was expected during conversion
    not_null,

    /// JSON bool was expected during conversion
    not_bool,

    /// JSON array was expected during conversion
    not_array,

    /// JSON object was expected during conversion
    not_object,

    /// JSON string was expected during conversion
    not_string,

    /// JSON array has size incompatible with target
    size_mismatch,

    /// none of the possible conversions were successful
    exhausted_variants,

    /// the key does not correspond to a known name
    unknown_name,
};

/** Error conditions corresponding to JSON errors
*/
enum class condition
{
    /// A parser-related error
    parse_error = 1,

    /// An error related to parsing JSON pointer string
    pointer_parse_error,

    /// An error related to applying JSON pointer string to a value
    pointer_use_error,

    /// A conversion error
    conversion_error,

    /// A generic error
    generic_error,
};

BOOST_JSON_NS_END

#include <boost/json/impl/error.hpp>

#endif
