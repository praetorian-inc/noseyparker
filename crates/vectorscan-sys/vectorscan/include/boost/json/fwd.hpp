//
// Copyright (c) 2019 Vinnie Falco (vinnie.falco@gmail.com)
// Copyright (c) 2022 Dmitry Arkhipov (grisumbras@yandex.ru)
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
//
// Official repository: https://github.com/boostorg/json
//

#ifndef BOOST_JSON_FWD_HPP
#define BOOST_JSON_FWD_HPP

#include <boost/json/detail/config.hpp>

BOOST_JSON_NS_BEGIN

// Forward declarations

#ifndef BOOST_JSON_DOCS

class array;
class object;
class string;
class value;

class key_value_pair;
class storage_ptr;

struct value_from_tag;

template<class T>
struct value_to_tag;

template<class T>
struct try_value_to_tag;

template<class T1, class T2>
struct result_for;

template<class T>
struct is_string_like;

template<class T>
struct is_sequence_like;

template<class T>
struct is_map_like;

template<class T>
struct is_tuple_like;

template<class T>
struct is_null_like;

template<class T>
struct is_described_class;

template<class T>
struct is_described_enum;

template<class T>
void value_from( T&& t, value& jv );

template<class T>
T value_to( value const & v );

template<class T>
typename result_for<T, value>::type
try_value_to( value const & jv );

template<class T>
typename result_for<T, value>::type
result_from_errno( int e, boost::source_location const* loc ) noexcept;

#endif

BOOST_JSON_NS_END

#endif
