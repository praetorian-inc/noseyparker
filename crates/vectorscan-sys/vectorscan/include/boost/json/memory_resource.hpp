//
// Copyright (c) 2019 Vinnie Falco (vinnie.falco@gmail.com)
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
//
// Official repository: https://github.com/boostorg/json
//

#ifndef BOOST_JSON_MEMORY_RESOURCE_HPP
#define BOOST_JSON_MEMORY_RESOURCE_HPP

#include <boost/json/detail/config.hpp>

#include <boost/container/pmr/memory_resource.hpp>
#include <boost/container/pmr/polymorphic_allocator.hpp>

BOOST_JSON_NS_BEGIN

#ifdef BOOST_JSON_DOCS

/** The type of memory resource used by the library.

    Alias for `boost::container::pmr::memory_resource`.
*/
class memory_resource
{
};

/** The type of polymorphic allocator used by the library.

    Alias template for `boost::container::pmr::polymorphic_allocator`.
*/
template<class T>
class polymorphic_allocator;

// VFALCO Bug: doc toolchain won't make this a ref
//using memory_resource = __see_below__;

#else
using memory_resource = boost::container::pmr::memory_resource;

template<class T>
using polymorphic_allocator =
    boost::container::pmr::polymorphic_allocator<T>;
#endif

/** Return true if a memory resource's deallocate function has no effect.

    This metafunction may be specialized to indicate to
    the library that calls to the `deallocate` function of
    a @ref memory_resource have no effect. The implementation
    will elide such calls when it is safe to do so. By default,
    the implementation assumes that all memory resources
    require a call to `deallocate` for each memory region
    obtained by calling `allocate`.

    @par Example

    This example specializes the metafuction for `my_resource`,
    to indicate that calls to deallocate have no effect:

    @code

    // Forward-declaration for a user-defined memory resource
    struct my_resource;

    // It is necessary to specialize the template from
    // inside the namespace in which it is declared:

    namespace boost {
    namespace json {

    template<>
    struct is_deallocate_trivial< my_resource >
    {
        static constexpr bool value = true;
    };

    } // namespace json
    } // namespace boost

    @endcode

    It is usually not necessary for users to check this trait.
    Instead, they can call @ref storage_ptr::is_deallocate_trivial
    to determine if the pointed-to memory resource has a trivial
    deallocate function.

    @see
        @ref memory_resource,
        @ref storage_ptr
*/
template<class T>
struct is_deallocate_trivial
{
    /** A bool equal to true if calls to `T::do_deallocate` have no effect.

        The primary template sets `value` to false.
    */
    static constexpr bool value = false;
};

BOOST_JSON_NS_END

#endif
