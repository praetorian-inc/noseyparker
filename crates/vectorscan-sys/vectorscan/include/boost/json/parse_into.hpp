//
// Copyright (c) 2021 Peter Dimov
// Copyright (c) 2021 Vinnie Falco (vinnie.falco@gmail.com)
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
//
// Official repository: https://github.com/boostorg/json
//

#ifndef BOOST_JSON_PARSE_INTO_HPP
#define BOOST_JSON_PARSE_INTO_HPP

#include <boost/json/detail/config.hpp>

#include <boost/json/basic_parser.hpp>
#include <boost/json/string_view.hpp>
#include <boost/json/system_error.hpp>
#include <boost/json/detail/parse_into.hpp>

namespace boost {
namespace json {

/** `basic_parser` that parses into a given type

    This is an alias template for @ref basic_parser instantiations that use
    a dedicated handler that parses directly into an object provided by the
    user instead of creating a @ref value.

    Objects of type `parser_for<T>` have constructor signature equivalent to
    `parser_for( parse_options const&, T& )`.

    @tparam T the type to parse into. This type must be
    [*DefaultConstructible*](https://en.cppreference.com/w/cpp/named_req/DefaultConstructible).
*/
template< class T >
using parser_for =
#ifndef BOOST_JSON_DOCS
    basic_parser<detail::into_handler<T>>;
#else
    __see_below__;
#endif

/** Parse a JSON text into a user-defined object.

    This function parses an entire string in one step and fills an object
    provided by the user. If the buffer does not contain a complete serialized
    JSON text, an error occurs. In this case `v` may be partially filled.

    The function supports default constructible types satisfying
    <a href="https://en.cppreference.com/w/cpp/named_req/SequenceContainer"><em>SequenceContainer</em></a>,
    arrays, arithmetic types, `bool`, `std::tuple`, `std::pair`,
    `std::optional`, `std::nullptr_t`, and structs and enums described using
    Boost.Describe.

    @par Complexity
    Linear in `sv.size()`.

    @par Exception Safety
    Basic guarantee.
    Calls to `memory_resource::allocate` may throw.

    @param v The type to parse into.

    @param sv The string to parse.

    @param ec Set to the error, if any occurred.

    @param opt The options for the parser. If this parameter
    is omitted, the parser will accept only standard JSON.
*/
/** @{ */
template<class V>
void
parse_into(
    V& v,
    string_view sv,
    error_code& ec,
    parse_options const& opt = {} );

template<class V>
void
parse_into(
    V& v,
    string_view sv,
    std::error_code& ec,
    parse_options const& opt = {} );
/** @} */

/** Parse a JSON text into a user-defined object.

    This function parses an entire string in one step and fills an object
    provided by the user. If the buffer does not contain a complete serialized
    JSON text, an exception is thrown. In this case `v` may be
    partially filled.

    The function supports default constructible types satisfying
    <a href="https://en.cppreference.com/w/cpp/named_req/SequenceContainer"><em>SequenceContainer</em></a>,
    arrays, arithmetic types, `bool`, `std::tuple`, `std::pair`,
    `std::optional`, `std::nullptr_t`, and structs and enums described using
    Boost.Describe.

    @par Complexity
    Linear in `sv.size()`.

    @par Exception Safety
    Basic guarantee.
    Throws @ref system_error on failed parse.
    Calls to `memory_resource::allocate` may throw.

    @param v The type to parse into.

    @param sv The string to parse.

    @param opt The options for the parser. If this parameter
    is omitted, the parser will accept only standard JSON.
*/
template<class V>
void
parse_into(
    V& v,
    string_view sv,
    parse_options const& opt = {} );

/** Parse a JSON text into a user-defined object.

    This function reads data from an input stream and fills an object provided
    by the user. If the buffer does not contain a complete serialized JSON
    text, or contains extra non-whitespace data, an error occurs. In this case
    `v` may be partially filled.

    The function supports default constructible types satisfying
    <a href="https://en.cppreference.com/w/cpp/named_req/SequenceContainer"><em>SequenceContainer</em></a>,
    arrays, arithmetic types, `bool`, `std::tuple`, `std::pair`,
    `std::optional`, `std::nullptr_t`, and structs and enums described using
    Boost.Describe.

    @par Complexity
    Linear in the size of consumed input.

    @par Exception Safety
    Basic guarantee.
    Calls to `memory_resource::allocate` may throw.
    The stream may throw as described by
    [`std::ios::exceptions`](https://en.cppreference.com/w/cpp/io/basic_ios/exceptions).

    @param v The type to parse into.

    @param is The stream to read from.

    @param ec Set to the error, if any occurred.

    @param opt The options for the parser. If this parameter
    is omitted, the parser will accept only standard JSON.
*/
/** @{ */
template<class V>
void
parse_into(
    V& v,
    std::istream& is,
    error_code& ec,
    parse_options const& opt = {} );

template<class V>
void
parse_into(
    V& v,
    std::istream& is,
    std::error_code& ec,
    parse_options const& opt = {} );
/** @} */

/** Parse a JSON text into a user-defined object.

    This function reads data from an input stream and fills an object provided
    by the user. If the buffer does not contain a complete serialized JSON
    text, or contains extra non-whitespace data, an exception is thrown. In
    this case `v` may be partially filled.

    The function supports default constructible types satisfying
    <a href="https://en.cppreference.com/w/cpp/named_req/SequenceContainer"><em>SequenceContainer</em></a>,
    arrays, arithmetic types, `bool`, `std::tuple`, `std::pair`,
    `std::optional`, `std::nullptr_t`, and structs and enums described using
    Boost.Describe.

    @par Complexity
    Linear in the size of consumed input.

    @par Exception Safety
    Basic guarantee.
    Throws @ref system_error on failed parse.
    Calls to `memory_resource::allocate` may throw.
    The stream may throw as described by
    [`std::ios::exceptions`](https://en.cppreference.com/w/cpp/io/basic_ios/exceptions).

    @param v The type to parse into.

    @param is The stream to read from.

    @param opt The options for the parser. If this parameter
    is omitted, the parser will accept only standard JSON.
*/
template<class V>
void
parse_into(
    V& v,
    std::istream& is,
    parse_options const& opt = {} );

} // namespace boost
} // namespace json

#include <boost/json/impl/parse_into.hpp>

#endif
