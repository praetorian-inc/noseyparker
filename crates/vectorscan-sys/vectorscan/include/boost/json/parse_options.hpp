//
// Copyright (c) 2020 Krystian Stasiowski (sdkrystian@gmail.com)
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
//
// Official repository: https://github.com/boostorg/json
//

#ifndef BOOST_JSON_PARSE_OPTIONS_HPP
#define BOOST_JSON_PARSE_OPTIONS_HPP

#include <boost/json/detail/config.hpp>
#include <iosfwd>

namespace boost {
namespace json {

/** Enumeration of number parsing modes

    These values are used to select the way to parse numbers.

    @see
        @ref parse_options,
        @ref basic_parser,
        @ref parser.
*/
enum class number_precision : unsigned char
{
    /// Fast, but potentially less precise mode.
    imprecise,

    /// Slower, but precise mode.
    precise,

    /// The fastest mode, that only validates encountered numbers without
    /// parsing them.
    none,
};

/** Parser options

    This structure is used for specifying
    maximum parsing depth, and whether
    to allow various non-standard extensions.
    Default-constructed options set maximum
    parsing depth to 32 and specify that only
    standard JSON is allowed,

    @see
        @ref basic_parser,
        @ref parser.
*/
struct parse_options
{
    /** Maximum nesting level of arrays and objects.

        This specifies the maximum number of nested
        structures allowed while parsing a JSON text. If
        this limit is exceeded during a parse, an
        error is returned.

        @see
            @ref basic_parser,
            @ref stream_parser.
    */
    std::size_t max_depth = 32;

    /** Number pasing mode

        This selects the way to parse numbers. The default is to parse them
        fast, but with possible slight imprecision for floating point numbers
        with larger mantissas. Users can also choose to parse numbers slower
        but with full precision. Or to not parse them at all, and only validate
        numbers. The latter mode is useful for @ref basic_parser instantiations
        that wish to treat numbers in a custom way.

        @see
            @ref basic_parser,
            @ref stream_parser.
    */
    number_precision numbers = number_precision::imprecise;

    /** Non-standard extension option

        Allow C and C++ style comments to appear
        anywhere that whitespace is permissible.

        @see
            @ref basic_parser,
            @ref stream_parser.
    */
    bool allow_comments = false;

    /** Non-standard extension option

        Allow a trailing comma to appear after
        the last element of any array or object.

        @see
            @ref basic_parser,
            @ref stream_parser.
    */
    bool allow_trailing_commas = false;

    /** Non-standard extension option

        Allow invalid UTF-8 sequnces to appear
        in keys and strings.

        @note This increases parsing performance.

        @see
            @ref basic_parser,
            @ref stream_parser.
    */
    bool allow_invalid_utf8 = false;

    /** Non-standard extension option

        Allow `Infinity`, `-Infinity`, and `NaN` JSON literals. These values
        are produced by some popular JSON implementations for positive
        infinity, negative infinity and NaN special numbers respectively.

        @see
            @ref basic_parser,
            @ref stream_parser.
    */
    bool allow_infinity_and_nan = false;

    /** Set JSON parse options on input stream.

        The function stores parse options in the private storage of the stream. If
        the stream fails to allocate necessary private storage, `badbit` will be
        set on it.

        @return Reference to `is`.

        @par Complexity
        Amortized constant (due to potential memory allocation by the stream).

        @par Exception Safety
        Strong guarantee.
        The stream may throw as configured by
        [`std::ios::exceptions`](https://en.cppreference.com/w/cpp/io/basic_ios/exceptions).

        @param is The input stream.

        @param opts The options to store.
    */
    BOOST_JSON_DECL
    friend
    std::istream&
    operator>>( std::istream& is, parse_options const& opts );
};

} // namespace json
} // namespace boost

#endif
