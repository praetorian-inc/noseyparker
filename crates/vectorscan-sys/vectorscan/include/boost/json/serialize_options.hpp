//
// Copyright (c) 2023 Dmitry Arkhipov (grisumbras@yandex.ru)
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
//
// Official repository: https://github.com/boostorg/json
//

#ifndef BOOST_JSON_SERIALIZE_OPTIONS_HPP
#define BOOST_JSON_SERIALIZE_OPTIONS_HPP

#include <boost/json/detail/config.hpp>
#include <iosfwd>

namespace boost {
namespace json {

/** Serialize options

    This structure is used for specifying whether to allow non-standard
    extensions. Default-constructed options specify that only standard JSON is
    produced.

    @see
        @ref serialize,
        @ref serializer.
*/
struct serialize_options
{
    /** Non-standard extension option

        Output `Infinity`, `-Infinity` and `NaN` for positive infinity,
        negative infinity, and "not a number" doubles.

        @see
            @ref serialize,
            @ref serializer.
    */
    bool allow_infinity_and_nan = false;

    /** Set JSON serialization options on input stream.
        The function stores serialization options in the private storage of the
        stream. If the stream fails to allocate necessary private storage,
        `badbit` will be set on it.

        @return Reference to `os`.

        @par Complexity
        Amortized constant (due to potential memory allocation by the stream).

        @par Exception Safety
        Strong guarantee.
        The stream may throw as configured by
        [`std::ios::exceptions`](https://en.cppreference.com/w/cpp/io/basic_ios/exceptions).

        @param os The output stream.

        @param opts The options to store.
    */
    BOOST_JSON_DECL
    friend
    std::ostream&
    operator<<( std::ostream& os, serialize_options const& opts );
};

} // namespace json
} // namespace boost

#endif // BOOST_JSON_SERIALIZE_OPTIONS_HPP
