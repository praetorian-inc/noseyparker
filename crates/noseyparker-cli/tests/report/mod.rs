use super::*;
pub use pretty_assertions::{assert_eq, assert_ne};

#[test]
fn report_nonexistent_default_datastore() {
    let scan_env = ScanEnv::new();
    let ds = scan_env.root.child("datastore.np");
    ds.assert(predicates::path::missing());

    assert_cmd_snapshot!(noseyparker!("report")
        .current_dir(scan_env.root.path())
        .assert()
        .failure());

    ds.assert(predicates::path::missing());
}

/// Test that the `report` command's `--max-matches` can be given a negative value (which means "no
/// limit" for the option) without requiring an equals sign for the value. That is, instead of
/// _requiring_ that the option be written `--max-matches=-1`, it should work fine to write
/// `--max-matches -1`.
///
/// N.B., Suppoorting that argument parsing requires passing the `allow_negative_numbers=true` in
/// the correct spot in the `clap` code.
#[test]
fn report_unlimited_matches() {
    let scan_env = ScanEnv::new();
    let input = scan_env.input_file_with_secret("input.txt");

    noseyparker_success!("scan", "-d", scan_env.dspath(), input.path())
        .stdout(match_scan_stats("104 B", 1, 1, 1));

    with_settings!({
        filters => get_report_stdout_filters(),
    }, {
        assert_cmd_snapshot!(noseyparker_success!("report", "-d", scan_env.dspath(), "--max-matches", "-1"));
    });
}

/// Test that the `report` command uses colors as expected when *not* running under a pty:
///
/// - When running with the output explicitly written to a file, colors are not used
///
/// - When running with with the output explicitly written to a file and `--color=always`
///   specified, colors are used
#[test]
fn report_output_colors1() {
    let scan_env = ScanEnv::new();
    let input = scan_env.input_file_with_secret("input.txt");

    let output1 = scan_env.child("findings.txt");
    let output2 = scan_env.child("findings.colored.txt");

    noseyparker_success!("scan", "-d", scan_env.dspath(), input.path())
        .stdout(match_scan_stats("104 B", 1, 1, 1));

    noseyparker_success!("report", "-d", scan_env.dspath(), "-o", output1.path());
    noseyparker_success!("report", "-d", scan_env.dspath(), "-o", output2.path(), "--color=always");

    let output1_contents = std::fs::read_to_string(output1.path()).unwrap();
    let output2_contents = std::fs::read_to_string(output2.path()).unwrap();

    assert_ne!(output1_contents, output2_contents);
    with_settings!({
        filters => get_report_stdout_filters(),
    }, {
        assert_snapshot!(output1_contents);
    });
    assert_eq!(&output1_contents, &console::strip_ansi_codes(&output2_contents));
}

fn read_json_file(fname: &Path) -> serde_json::Value {
    let file = std::fs::File::open(fname).unwrap();
    let mut reader = std::io::BufReader::new(file);
    let findings = serde_json::from_reader(&mut reader).unwrap();
    findings
}

/// Test that the `report --finding-status` option works as expected.
/// In the case of a newly-created datastore, there will be no statuses assigned at all, so we do
/// only some basic checks.
#[test]
fn report_finding_status() {
    use serde_json::json;

    let scan_env = ScanEnv::new();
    let input = scan_env.input_file_with_secret("input.txt");
    noseyparker_success!("scan", "-d", scan_env.dspath(), input.path())
        .stdout(match_scan_stats("104 B", 1, 1, 1));

    let report = |out: &ChildPath, status: &str| {
        noseyparker_success!(
            "report",
            "-d",
            scan_env.dspath(),
            "--format=json",
            "-o",
            out.path(),
            "--finding-status",
            status
        );
    };

    // case 1: accept
    let output = scan_env.child("findings.accept.json");
    report(&output, "accept");
    let findings = read_json_file(output.path());
    assert_eq!(findings, json!([]));

    // case 2: reject
    let output = scan_env.child("findings.reject.json");
    report(&output, "reject");
    let findings = read_json_file(output.path());
    assert_eq!(findings, json!([]));

    // case 3: mixed
    let output = scan_env.child("findings.mixed.json");
    report(&output, "mixed");
    let findings = read_json_file(output.path());
    assert_eq!(findings, json!([]));

    // case 4: null
    let output = scan_env.child("findings.null.json");
    report(&output, "null");
    let findings = read_json_file(output.path());
    assert!(findings.is_array());
    assert_eq!(findings.as_array().unwrap().len(), 1);
}

// Test that the `report` command uses colors as expected when running under a pty:
// - When running with the output going to stdout (default), colors are used
// - When running with the explicitly written to a file, colors are not used
// XXX to get a pty, look at the `pty-process` crate: https://docs.rs/pty-process/latest/pty_process/blocking/struct.Command.html
