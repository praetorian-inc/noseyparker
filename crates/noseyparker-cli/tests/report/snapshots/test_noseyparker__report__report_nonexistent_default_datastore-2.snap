---
source: crates/noseyparker-cli/tests/report/mod.rs
expression: stdout
---

